File: mice.lua
local ffi = require'ffi'
local mice = ffi.load'./libmice.so'

ffi.cdef [[

struct input_event {
uint8_t time[8];
uint16_t type;
uint16_t code;
int32_t value;
};

int open_mice(void);
void close_mice(int fd);
struct input_event* read_event(int fd);

]]

local fd = mice.open_mice()
while (true)
do
  evt = mice.read_event(fd)
  print('type:'..evt.type..',code:'..evt.code..',val:'..evt.value)
  if ( evt.value ~= 589826 ) then break end
end

mice.close_mice(fd)

File: mice.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <linux/input.h>
#include <time.h>
#include <termios.h>
#include <linux/uinput.h>

static struct input_event evt;
static int fd;

int open_mice(void)
{
  fd = open("/dev/input/event0", O_RDWR );
  return fd;
}

void close_mice(int fd)
{
  close(fd);
}

struct input_event* read_event(int fd_lua)
{
  if ( read(fd, &evt, sizeof(evt)) < 0 )
    perror("can't read mice");
  return &evt;
}

